File: leetcode/longest_substring_without_repeating_characters.cpp
#include <iostream>
#include <unordered_map>
#include <utility>
#include <string>

using namespace std;

class Solution {
    public:
        int lengthOfLongestSubstring(string s) {
            unsigned longest = 0;

            char c;
            unordered_map<char, int> *seen = new unordered_map<char, int>;
            for (unsigned i=0; i<s.size(); i++) {
                c = s[i];
                if (seen->find(c) == seen->end()) {
                    seen->insert( make_pair(c, i) );
                } else {
                    i = seen->at(c);
                    longest = max(longest, (unsigned)seen->size());
                    delete seen;
                    seen = new unordered_map<char, int>;
                }
            }

            longest = max(longest, (unsigned)seen->size());
            delete seen;

            return longest;
        }
};

int main() {
    Solution S;

    string q = "dvdf";

    cout << S.lengthOfLongestSubstring(q) << endl;

    return 0;
}

File: leetcode/longest_substring_without_repeating_characters.py
def lengthOfLongestSubstring(s: str) -> int:
    max_len = 0
    last_seen = {c: -1 for c in s}

    last_repeat = -1
    for i, c in enumerate(s):
        print(s[:i+1], last_seen, last_repeat, max_len)
        max_len = max(max_len, i - max(last_seen[c], last_repeat))

        if c in last_seen:
            last_repeat = max(last_repeat, last_seen[c])

        last_seen[c] = i

    return max_len


print(lengthOfLongestSubstring('abcabcbb'))
print(lengthOfLongestSubstring('bbbbb'))
print(lengthOfLongestSubstring('pwwkew'))
print(lengthOfLongestSubstring('dvdf'))
print(lengthOfLongestSubstring('au'))
